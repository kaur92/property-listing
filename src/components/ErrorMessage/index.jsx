import React from "react";
import "./ErrorMessage.css";

const ErrorMessage = () => (
  <div className="error-page">
    <h1>Oops!!</h1>
    <h3>Something went wrong please try again.</h3>
  </div>
);

export default ErrorMessage;
