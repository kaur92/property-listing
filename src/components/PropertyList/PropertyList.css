.cards {
  max-width: 1200px;
}

.title {
  padding: 10px;
}
