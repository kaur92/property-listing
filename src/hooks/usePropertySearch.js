import { useEffect, useState } from "react";

const BASE_URI = "http://localhost:3001";

export const usePropertySearch = () => {
  const [isLoading, setIsLoading] = useState(false);
  const [properties, setProperties] = useState([]);
  const [error, setError] = useState(null);

  const fetchProperties = async () => {
    try {
      const response = await fetch(`${BASE_URI}/properties`);
      const data = await response.json();
      setProperties(data);
      setIsLoading(false);
    } catch (error) {
      setError(error);
      setIsLoading(false);
    }
  };

  useEffect(() => {
    setIsLoading(true);
    setProperties([]);
    fetchProperties();
  }, []);

  return { isLoading, properties, error };
};
