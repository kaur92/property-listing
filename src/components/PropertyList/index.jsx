import React from "react";
import { usePropertySearch } from "../../hooks/usePropertySearch";
import LoadingSpinner from "../LoadingSpinner";
import PropertyCard from "../PropertyCard";
import ErrorMessage from "../ErrorMessage";
import "./PropertyList.css";

const PropertyList = () => {
  const { isLoading, properties, serverError } = usePropertySearch();
  const renderProperty = (
    <>
      {serverError && <ErrorMessage />}
      {properties.length && (
        <>
          <div className="title">
            <h3>Properties you may like</h3>
          </div>
          <section className="cards">
            {properties.map((property) => (
              <PropertyCard key={property.id} property={property} />
            ))}
          </section>
        </>
      )}
      {!properties.length && (
        <div>
          <div className="title">
            <h3>No property found</h3>
          </div>
        </div>
      )}
    </>
  );

  return (isLoading ? <LoadingSpinner /> : renderProperty);
};

export default PropertyList;
