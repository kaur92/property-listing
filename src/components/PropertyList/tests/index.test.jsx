import renderer from "react-test-renderer";
import PropertyList from "../index";
import * as hooks from "../../../hooks/usePropertySearch";
import ErrorMessage from "../../ErrorMessage";

describe("<PropertyList /> tests", () => {
  describe("Empty property list", () => {
    it("should render correctly", async () => {
      jest.spyOn(hooks, "usePropertySearch").mockImplementation(() => ({
        properties: [],
        loading: false,
        serverError: null,
      }));

      const listComponent = renderer.create(<PropertyList />).toJSON();
      expect(listComponent).toMatchSnapshot();
    });
  });

  describe("Loading true", () => {
    it("should render correctly", async () => {
      jest.spyOn(hooks, "usePropertySearch").mockImplementation(() => ({
        properties: [],
        loading: true,
        serverError: null,
      }));

      const listComponent = renderer.create(<PropertyList />).toJSON();
      expect(listComponent).toMatchSnapshot();
    });
  });

  describe("Return error", () => {
    it("should render correctly", async () => {
      jest.spyOn(hooks, "usePropertySearch").mockImplementation(() => ({
        properties: [],
        loading: true,
        serverError: "error",
      }));

      const listComponent = renderer.create(<PropertyList />).toJSON();
      expect(listComponent).toMatchSnapshot();
    });
  });

  describe("Two properties in list", () => {
    it("should render correctly", async () => {
      jest.spyOn(hooks, "usePropertySearch").mockImplementation(() => ({
        properties: [
          {
            id: "bb32dc153a844c01a051ab48fa543c6f",
            address: "Gloucester Road, London SW7 4RT",
            image:
              "https://lid.zoocdn.com/u/1024/768/2b3d21c22e792b3093479bed79b84b931e4cc062.jpg:p",
            askingPrice: 50000,
            status: true,
            numberOfBedrooms: 1,
            propertyType: "flat",
          },
          {
            id: "bb32dc153a844c01a051ab48fa544r9y",
            address: "Princes Gate, London SW7 9FT",
            image:
              "https://lid.zoocdn.com/u/1024/768/c0ecdf355a66d5c571fe35773084cb7a64dc2907.jpg:p",
            askingPrice: 24750000,
            status: true,
            numberOfBedrooms: 7,
            propertyType: "flat",
          },
        ],
        loading: false,
        serverError: null,
      }));

      const listComponent = renderer.create(<PropertyList />).toJSON();
      expect(listComponent).toMatchSnapshot();
    });
  });
});
