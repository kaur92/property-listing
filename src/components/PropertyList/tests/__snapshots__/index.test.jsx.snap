// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PropertyList /> tests Empty property list should render correctly 1`] = `
Array [
  "0",
  <div>
    <div
      className="title"
    >
      <h3>
        No property found
      </h3>
    </div>
  </div>,
]
`;

exports[`<PropertyList /> tests Loading true should render correctly 1`] = `
Array [
  "0",
  <div>
    <div
      className="title"
    >
      <h3>
        No property found
      </h3>
    </div>
  </div>,
]
`;

exports[`<PropertyList /> tests Return error should render correctly 1`] = `
Array [
  <div
    className="error-page"
  >
    <h1>
      Oops!!
    </h1>
    <h3>
      Something went wrong please try again.
    </h3>
  </div>,
  "0",
  <div>
    <div
      className="title"
    >
      <h3>
        No property found
      </h3>
    </div>
  </div>,
]
`;

exports[`<PropertyList /> tests Two properties in list should render correctly 1`] = `
Array [
  <div
    className="title"
  >
    <h3>
      Properties you may like
    </h3>
  </div>,
  <section
    className="cards"
  >
    <div>
      <div
        className="card"
      >
        <div
          className="card-image"
        >
          <img
            alt=""
            src="https://lid.zoocdn.com/u/1024/768/2b3d21c22e792b3093479bed79b84b931e4cc062.jpg:p"
          />
          <div
            className="status status-active"
          >
            Active
          </div>
        </div>
        <div
          className="card-details"
        >
          <div
            className="price"
          >
            <label>
              Asking Price
            </label>
            <span>
              £
              50,000
            </span>
          </div>
          <div
            className="bathroom"
          >
            <label>
              1 bed flat for sale
            </label>
          </div>
          <div
            className="address"
          >
            <p>
              Gloucester Road, London SW7 4RT
            </p>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="card"
      >
        <div
          className="card-image"
        >
          <img
            alt=""
            src="https://lid.zoocdn.com/u/1024/768/c0ecdf355a66d5c571fe35773084cb7a64dc2907.jpg:p"
          />
          <div
            className="status status-active"
          >
            Active
          </div>
        </div>
        <div
          className="card-details"
        >
          <div
            className="price"
          >
            <label>
              Asking Price
            </label>
            <span>
              £
              24,750,000
            </span>
          </div>
          <div
            className="bathroom"
          >
            <label>
              7 bed flat for sale
            </label>
          </div>
          <div
            className="address"
          >
            <p>
              Princes Gate, London SW7 9FT
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>,
]
`;
